src/link/script.y

```diff
@@ -0,0 +1,479 @@
+%language "c++"
+%define api.value.type variant
+%define api.token.constructor
+
+%code requires {
+	#include <stdint.h>
+	#include <string>
+
+	#include "linkdefs.hpp"
+}
+%code {
+	#include <algorithm>
+	#include <array>
+	#include <assert.h>
+	#include <bit>
+	#include <cinttypes>
+	#include <fstream>
+	#include <locale>
+	#include <string_view>
+	#include <vector>
+
+	#include "helpers.hpp"
+	#include "itertools.hpp"
+	#include "util.hpp"
+
+	#include "link/main.hpp"
+	#include "link/section.hpp"
+
+	using namespace std::literals;
+
+	static void includeFile(std::string &&path);
+	static void incLineNo(void);
+
+	static void setSectionType(SectionType type);
+	static void setSectionType(SectionType type, uint32_t bank);
+	static void setAddr(uint32_t addr);
+	static void alignTo(uint32_t alignment, uint32_t offset);
+	static void pad(uint32_t length);
+	static void placeSection(std::string const &name);
+
+	static yy::parser::symbol_type yylex(void);
+
+	struct Keyword {
+		std::string_view name;
+		yy::parser::symbol_type (* tokenGen)(void);
+	};
+}
+
+%token YYEOF 0
+%token newline
+%token ORG "ORG"
+       INCLUDE "INCLUDE"
+       ALIGN "ALIGN"
+       DS "DS"
+%code {
+	static std::array keywords{
+		Keyword{"ORG"sv,     yy::parser::make_ORG},
+		Keyword{"INCLUDE"sv, yy::parser::make_INCLUDE},
+		Keyword{"ALIGN"sv,   yy::parser::make_ALIGN},
+		Keyword{"DS"sv,      yy::parser::make_DS},
+	};
+}
+%token <std::string> string;
+%token <uint32_t> number;
+%token <SectionType> section_type;
+
+%%
+
+lines: %empty
+     | line lines
+;
+
+line: INCLUDE string newline { includeFile(std::move($2)); } // Note: this additionally increments the line number!
+    | directive newline { incLineNo(); }
+    | newline { incLineNo(); }
+    | error newline { yyerrok; incLineNo(); } // Error recovery.
+;
+
+directive: section_type { setSectionType($1); }
+         | section_type number { setSectionType($1, $2); }
+         | ORG number { setAddr($2); }
+         | ALIGN number { alignTo($2, 0); }
+         | DS number { pad($2); }
+         | string { placeSection($1); }
+;
+
+%%
+
+#define scriptError(context, fmt, ...) ::error(NULL, 0, "%s(%" PRIu32 "): " fmt, \
+                                               context.path.c_str(), context.lineNo, __VA_ARGS__)
+// MSVC doesn't support __VA_OPT__ yet.
```

Comment:
Is `##` too nonstandard? MSVC supports it, and before that, would elide commas even without `##`: <https://learn.microsoft.com/en-us/cpp/preprocessor/preprocessor-experimental-overview?view=msvc-170#comma-elision-in-variadic-macros>

Reply:
Ah? I thought MSVC didn't support it. But if it does (under its default configuration, hopefully...), then we may look into that.

Reply:
Oh, GitHub Actions has MSVC 16.11, so we should be able to use `/Zc:preprocessor`! That seems stable, unlike `/experimental:preprocessor`. I'll try that out!

Reply:
Yep, fixed by a44df325ae24388ffcccc0a448651c578d88d77e. Thanks!